### PR TITLE
Java with multipass

### PR DESCRIPTION
**Pull request by jon-morra-zefr**
Previously the JNI interface did not support that ability to do multiple passes over the data.  This PR fixes that.  In addition this PR exposes the base Java abstract class `VWLearnerBase` so we can avoid having a typed VW learned when just doing learning.

### CODE REVIEW

File: java/src/main/java/vowpalWabbit/learner/VWLearnerBase.java [hunk 1]
```diff
@@ -17,7 +17,7 @@
  *
  * @author deak
  */
-abstract class VWLearnerBase<T> extends VWBase implements VWTypedLearner<T> {
+public abstract class VWLearnerBase<T> extends VWBase implements VWTypedLearner<T> {
```
Comment (ryan-deak-zefr) [inline java/src/main/java/vowpalWabbit/learner/VWLearnerBase.java]: This is a bad idea I think.  `VWTypedLearner<T>` has everything we need: `predict` and `learn` methods and `close`.  When calling `VWLearners.create`,  just do `VWLearners.create<Object>` or something.

File: java/src/main/java/vowpalWabbit/learner/VWLearners.java [hunk 2]
```diff
@@ -1,17 +1,12 @@
...
-import java.util.concurrent.locks.ReentrantLock;
-
 /**
  * This is the only entrance point to create a VWLearner.  It is the responsibility of the user to supply the type they want
  * given the VW command.  If that type is incorrect a {@link java.lang.ClassCastException} is thrown.  Refer to
  * {@link #create(String)} for more information.
  * @author jmorra
  */
 final public class VWLearners {
-    private static final Lock STATIC_LOCK = new ReentrantLock();
```
Comment (ryan-deak-zefr) [inline java/src/main/java/vowpalWabbit/learner/VWLearners.java]: Yup.  Should be removed.

File: java/src/main/java/vowpalWabbit/learner/VWLearnerBase.java (hunk 1 above)
Comment (JohnLangford) [inline java/src/main/java/vowpalWabbit/learner/VWLearnerBase.java]: @jon-morra-zefr what is the use case for this?

Comment (jon-morra-zefr) [inline java/src/main/java/vowpalWabbit/learner/VWLearnerBase.java]: We don't need it, it turns out that @ryan-deak-zefr is correct.

**Comment by JohnLangford** [comment]
Merged, thanks :-)

**Pull request merged**